### ISSUE TITLE
Observing difference in outputs from decoder with IO bindings.

### ISSUE THREAD

**Issue by VikasOjha666** [Ki6an/fastT5#49]
Hi @Ki6an  Was trying to implement IO bindings for the decoder part of the model. Used the same code from your repo to convert the model to ONNX. After loading the model and making predictions using the decoder session directly the output appears to be fine but with inputs binded the result is coming to be different. 

Below is the code for the IO bindings:
```python
def dec_pred_with_io_bindings(input_ids, attention_mask, encoder_output, past_key_values_dict,dec_session):
  dec_io_binding = dec_session.io_binding()
  dec_io_binding.bind_input(name="input_ids",
                          device_type="cuda",
                          device_id=0,
                          element_type=np.longlong,
                          shape=list(input_ids.shape),
                          buffer_ptr=input_ids.data_ptr())
  dec_io_binding.bind_input(name="encoder_attention_mask",
                          device_type="cuda",
                          device_id=0,
                          element_type=np.longlong,
                          shape=list(attention_mask.shape),
                          buffer_ptr=attention_mask.data_ptr())
                        
  dec_io_binding.bind_input(name="encoder_hidden_states",
                          device_type="cuda",
                          device_id=0,
                          element_type=np.float32,
                          shape=list(encoder_output.shape),
                          buffer_ptr=encoder_output.data_ptr())
  

  for key,val in past_key_values_dict.items():
    dec_io_binding.bind_input(name=key,
                                      device_type="cuda",
                                      device_id=0,
                                      element_type=np.float32,
                                      shape=list(val.shape),
                                      buffer_ptr=val.data_ptr())
  
  #Bind outputs.
  for arg in self.decoder.get_outputs():
    dec_io_binding.bind_output(arg.name, "cuda")
    
  dec_session.run_with_iobinding(dec_io_binding)
  ort_output = dec_io_binding.get_outputs()

  logits=ort_output[0]

  list_pkv = tuple(torch.from_numpy(x.numpy()).cuda() for x in ort_output[1:])

  # creates a tuple of tuples of shape 6x4 from the above tuple
  out_past_key_values = tuple(
      list_pkv[i : i + 4] for i in range(0, len(list_pkv), 4)
  )


  return torch.from_numpy(logits.numpy()).cuda(),out_past_key_values
```

**Comment by Ki6an** [comment]
could you give an example of how the results are different?

**Comment by VikasOjha666** [comment]
@Ki6an Thanks for taking time for your reply. When I pass the same decoder_input_ids, attention_mask,encoder_outputs and past key values to the prediction function then thr resulting logits and past key values are coming to be different.

**Comment by VikasOjha666** [comment]
@Ki6an  Please take a look at this https://drive.google.com/file/d/1EWHP7XzdX45MwHHPnFcExp55Vw_DahTX/view?usp=sharing notebook which demonstrates the prediction on colab. I have saved the input to the model in  a pickle file which can be found at this link https://drive.google.com/file/d/1KTQviIEmS06y9OpFRyfznUTJXhHP3VQ0/view?usp=sharing .The notebook already contains the comparison of the outputs.

**Comment by Ki6an** [comment]
from the initial look, I can say that `torch.from_numpy(x.numpy()).cuda()` is slow.. possible [workaround ](https://github.com/microsoft/onnxruntime/issues/10327)

**Comment by VikasOjha666** [comment]
@Ki6an  Thanks for your response. I highly appreciate it. I will try it and let you know.

**Comment by VikasOjha666** [comment]
@Ki6an But also I have the question that in the function above the conversion is taking after prediction right?. In another code I didn't do the conversion and directly compared the ORT values still the output was different hence it isn't something else?

**Comment by VikasOjha666** [comment]
@Ki6an In the notebook I provided. The past_key_values have already been flattened and converted to torch Cuda tensors.  It's a dictionary in this format. {'pkv_0':past_key_val_0,'pkv_1':past_key_val_1......} hence the conversion is not taking placing during prediction as it already have been converted.

**Comment by VikasOjha666** [comment]
@Ki6an  I have one more update. When I converted all the inputs to ORT values and ran with run_with_ort_values of the ONNX session then it produced the correct output. Below is the code for that:

decoder_input_ids=model_inputs['decoder_input_ids']
decoder_input_ids=onnxruntime.OrtValue.ortvalue_from_numpy(decoder_input_ids.cpu().numpy(),"cuda")
attention_mask_val=onnxruntime.OrtValue.ortvalue_from_numpy(model_inputs['attention_mask'].cpu().numpy(),"cuda")
flat_past_key_values = functools.reduce(operator.iconcat, model_inputs['past_key_values'], [])
past_key_values_ort = {
    f"pkv_{i}": onnxruntime.OrtValue.ortvalue_from_numpy(pkv.cpu().numpy(),"cuda") for i, pkv in enumerate(flat_past_key_values)
}
encoder_outputs_ort=onnxruntime.OrtValue.ortvalue_from_numpy(model_inputs['encoder_outputs'].last_hidden_state.cpu().numpy(),"cuda")
decoder_inputs_ort = {
            "input_ids": decoder_input_ids,
            "encoder_attention_mask": attention_mask_val,
            "encoder_hidden_states":encoder_outputs_ort,
        }

outs_ort=dec_session.run_with_ort_values(None,{**decoder_inputs_ort, **past_key_values_ort})

**Comment by Ki6an** [comment]
for io_binding this should fix it https://github.com/microsoft/onnxruntime/issues/10992

**Comment by VikasOjha666** [comment]
@Ki6an Thanks for your time and the answer. By the way binding ORT values instead of tensors also solves the issue.But this is clearly efficient.
